### ISSUE TITLE
createGlobalStyle creates memory leak in Next.js

### ISSUE THREAD

**Issue by robbue** [styled-components/styled-components#3022]
### Binaries:
 - Node: 13.7.0 - ~/.nvm/versions/node/v13.7.0/bin/node
 - Yarn: 1.21.1 - /usr/local/bin/yarn
 - npm: 6.13.6 - ~/.nvm/versions/node/v13.7.0/bin/npm
### npmPackages:
 - babel-plugin-styled-components: ^1.10.7 => 1.10.6 
 - styled-components: ^5.0.1 => 5.0.1 

## Problem
We are using styled-components and Next.js (7.2.1) and we have experienced memory leaks in our app for some time, so I started to investigate it, in the process I found other issues here about: 
- using the css helper when composing styling partials to be interpolated into a styled component
- not to use `@import` inside `createGlobalStyle`
- not dynamically create components

So I went through the whole app making sure to avoid these. I deployed a new version but the memory leak was still present. As seen on the graph below:
<img width="942" alt="Skjermbilde 2020-02-14 19 49 28" src="https://user-images.githubusercontent.com/1470041/74558893-6a2e5780-4f63-11ea-99b6-ceb72fed1239.png">

After inspecting the node server I found out it was `createGlobalStyle` in our [`_app.js`](https://nextjs.org/docs/advanced-features/custom-app) (Next.js file) that caused the memory to rise. In the `createGlobalStyle` we used `fontFace`/`normalize` from [polished](https://github.com/styled-components/polished)), and around 50 other lines of reset css (nothing special).

Removing `createGlobalStyle` solved the memory leak  (or at least reduced to a minimum). 
See the graph below:
<img width="951" alt="Skjermbilde 2020-02-14 19 50 55" src="https://user-images.githubusercontent.com/1470041/74558899-6bf81b00-4f63-11ea-9d33-ccc43076c023.png">

Did we use `createGlobalStyle`wrong or do someone have some insight to this?

**Comment by boyuan459** [comment]
in our next.js project, we can only use one createGlobalStyle, other global styles is replaced with next.js <style global jsx>

**Comment by PaulACoroneos** [comment]
Yes we are having the same issue in our Next app as well. Only difference is that we are using latest Next version.

**Comment by PaulACoroneos** [comment]
We confirmed the rollback resolved the issue. Our current style file is just a copy of normalize css and reset css with some font-face definitions. Something like

export const AppGlobalStyle = createGlobalStyle<IAssetsProps>`
  ${reset}
  ${normalize}

  @font-face { 
...
`;

Where reset and normalize are something like 

const normalize = `
html {
  line-height: 1.15;
}
....
`

**Comment by PaulACoroneos** [comment]
One thing I noticed today. We have a block comment in our globalstyle. I wonder if this is possibly related to #3018

**Comment by robbue** [comment]
By removing `createGlobalStyle` the memory leak dropped significantly, but we still experience a leak (after 02/14). It went from 7 to 43 % memory utilization over 2 weeks, and we can live with that for now. But will try to investigate further.

<img width="714" alt="Skjermbilde 2020-02-29 11 59 25" src="https://user-images.githubusercontent.com/1470041/75606245-18510a00-5aeb-11ea-8ef0-4afd5dc035f5.png">

**Comment by ghengeveld** [comment]
We are running StoryShots over our 2000+ stories and after upgrading to Styled Components v5 our tests are running out. I've been able to bring down the test time drastically by stripping our global styles (with `createGlobalStyle`) down to just a few simple css rules. At one point adding a single rule on an element selector increased test time from 12 minutes to 17 minutes. Adding more increased it to 55 minutes, and that wasn't even our complete set of global styles. Before upgrading, our tests took only 3 minutes or so.

**Comment by joshamos** [comment]
We just had to revert a PR that added our first call to `createGlobalStyle` due to a memory leak.

![image](https://user-images.githubusercontent.com/131831/75821375-3cb22c80-5df2-11ea-83ce-def33b039861.png)

Using `styled-components@5.0.0` with a custom SSR set up (not NextJS).

The contents of the css in `createGlobalStyle` was pretty gnarly, and old sass based CSS reset - copied somewhat verbatim - comments and all.

**Comment by PaulACoroneos** [comment]
@joshamos  Yes this is exactly what happened to us.

**Comment by probablyup** [comment]
@ghengeveld are you still seeing that with 5.0.1? 5.0.0 had a subtle bug in it relating to SSR that could cause some weird lag.

We also have some unreleased changes in `styled-components@test` if y'all could give it a look.

**Comment by wwwebman** [comment]
@ghengeveld The issue is still there :( 

**Packages versions:**
"styled-components": "^5.0.1"
"next": "9.2.2"

**One of our global styles:**
```
import { createGlobalStyle } from 'styled-components';

const AppGlobalStyles = createGlobalStyle`
  #__next {
     display: flex;
     flex-direction: column;
     height: 100vh;
  }
`;

AppGlobalStyles.displayName = 'AppGlobalStyles';

export default AppGlobalStyles;
```

**Memory usage  with `createGlobalStyle` ^^:**
![Screenshot from 2020-03-06 14-39-07](https://user-images.githubusercontent.com/13425833/76088593-906e7280-5fb8-11ea-80ec-89a9419814a7.png)

**Memory usage  without `createGlobalStyle`:**
![Screenshot from 2020-03-06 14-40-19](https://user-images.githubusercontent.com/13425833/76088694-c1e73e00-5fb8-11ea-9f8f-cfaab0f441ef.png)

For sure we have some other small memory leaks, but `createGlobalStyle` produces the biggest one. This is how our memory usage looks using  `createGlobalStyle`:

![Screenshot from 2020-03-06 14-35-34](https://user-images.githubusercontent.com/13425833/76088828-007cf880-5fb9-11ea-8441-c8d8629aea1e.png)

**Constructor:**
`#__next{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-flex-direction:column;-ms-flex-direction:column;flex-direction:column;height:100vh;/*\|*/}`:

![Screenshot from 2020-03-06 14-46-53](https://user-images.githubusercontent.com/13425833/76089069-5a7dbe00-5fb9-11ea-9b8c-3ddb15d9aae6.png)

**The object  referred to the constructor ^^:**
![Screenshot from 2020-03-06 14-47-45](https://user-images.githubusercontent.com/13425833/76089263-a3357700-5fb9-11ea-89f6-da4e7fa79599.png)

As you can see the memory is being actively spammed on each page render by CSS string produced by  `createGlobalStyle` output. This is a huge issue. Possible solutions:
- Downgrade as mentioned by @ghengeveld 
- `import '../global.css'`
-  Fix `createGlobalStyle` ASAP

**Comment by GLObus303** [comment]
We were also forced to downgrade to 4.4.3. This is our memory consumption before and after styled-components v5, I am sure you can pinpoint the points where the version changed :)).

![Uploading image.png…]()

**Comment by alexparish** [comment]
I have been suffering from a memory leak in my Next.js app and after plenty of searching came across this issue which appears to be the cause of my issue.

I am using `createGlobalStyle` and `styled-normalize` in `_app.js` and can see you the strings in the Chrome DevTools Heap Snapshots:

<img width="1193" alt="Screen Shot 2020-04-12 at 15 44 05" src="https://user-images.githubusercontent.com/135674/79071595-7c631280-7cd4-11ea-849d-aace2417bd2a.png">

*Package versions:*
"styled-components": "5.0.1" / "5.1.0" 
"next": "9.3.1"

You can see my `_app.js` here: https://gist.github.com/alexparish/ad63a7822884be77e16a093bfb5fd992 

I'd be happy to create a repo if the maintainers would find it useful to reproduce the issue?

**Comment by alexparish** [comment]
Has anyone been able to resolve this?

I wonder if the issue is specifically with Next.js rather than styled-components. For example would any large string located in `_app.js` cause a memory leak?

**Comment by peeter-tomberg** [comment]
This is not related to Next.js. We're experiencing the same problems with a custom built React SSR implementation. Downgrading to version 4 solved the issue.

Seeing the same picture in my debugger as https://github.com/styled-components/styled-components/issues/3022#issuecomment-612628550

**Comment by gregorskii** [comment]
We ran into a similar issue where stylis `trim` function was causing a mem leak. Downgrading to 4.4.0, validating on server.

**Comment by Oikio** [comment]
Same problem on SSR with Next.js and Styled-components v5.1.1. Could not downgrade, so workaround was just not to use `createGlobalStyle`, which causes it.

**Comment by alexparish** [comment]
> Same problem on SSR with Next.js and Styled-components v5.1.1. Could not downgrade, so workaround was just not to use `createGlobalStyle`, which causes it.

Out of interest what was your workaround? An alternative approach to using `createGlobalStyle`?

**Comment by Oikio** [comment]
Just a string in style tag, put in head:
```
export const GlobalStyles: any = () => (
  <style
    dangerouslySetInnerHTML={{
      __html: `
 ... styles ...
`
    }}
  />
)
```
Using Typescript, so had to cast an any type to use this component as before

**Comment by probablyup** [comment]
Need a reproduction

**Comment by gregorskii** [comment]
I think it's going to be quite hard to reproduce without access to one of our hosting environments. Can you check my heap dump photo above? That could elude to the issue.

**Comment by Oikio** [comment]
I have tried to create a bare bones (no frameworks) reproduction case, but I could not reproduce memory leak. Maybe I'm doing something wrong, so someone can take a look and update this repo: https://github.com/Oikio/styled-components-memory-leak-reproduction.

On our production we have Nextjs and lots of things happening of course, but I could see allocation of strings from createGlobalStyle growing. If I will find time later, I will try to create reproduction with Nextjs.

**Comment by ellioseven** [comment]
We ran into this issue on production. Like most others indicated, this was due to `createGlobalStyle` creating many strings on each page visit. Like others, we had to revert to v4.

Versions:

```
next@8.1.0
styled-components@5.2.0
babel-plugin-styled-components@1.10.0
```

Will update when I have a reproduction available.

**Comment by Oikio** [comment]
I'm not common with the code base, but investigating it further led to Sheet.js, this particular line:
```
// line 69
  allocateGSInstance(id: string) {
    return (this.gs[id] = (this.gs[id] || 0) + 1);
  }
```

New style sheets are allocated later with this unique number in `const styleSheet = useStyleSheet();`, polluting the `names` map entry:
<img width="329" alt="image" src="https://user-images.githubusercontent.com/1791802/96729145-2246b680-13b5-11eb-928d-be1cf4bb8664.png">

Why this happening and is the reasoning behind incrementing in `allocateGSInstance` I don't know. Will investigate further, but it would be great if someone with better styled-components project knowledge could take a look at it.

**Comment by probablyup** [comment]
@Oikio what does your global style look like, are there a lot of dynamic interpolations?

**Comment by probablyup** [comment]
`this.gs` should be isolated to the `ServerStyleSheet` created on each SSR pass (it is not meant to be reused), so it's odd that it would have unbound growth like that unless you have a ton of global styles being injected all over the place

**Comment by Oikio** [comment]
@probablyup  only 3 styles, they all look similar, not much happening there, I've even tried to remove all code from them, so that it would just a string passed to `createGlobalStyle`, but result was the same, here is an example of styles I have:

```
import { createGlobalStyle } from 'styled-components'
import { getColor } from './tokenGetter'

export const GlobalStyleReset = createGlobalStyle`
html {
  -webkit-text-size-adjust: 100%;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -webkit-tap-highlight-color: transparent;
}

body {
  margin: 0;
  line-height: 1;
  font-family: sans-serif;
  background-color: ${getColor('background')};
}

iframe {
  border: 0;
}

ul,
ol {
  margin-top: 0;
  margin-bottom: 0;
  padding-left: 0;
}

li {
  display: block;
}

blockquote {
  margin: 0;
  padding: 0;
}

h1,
h2,
h3,
h4,
h5,
h6,
p {
  margin: 0;
  padding: 0;
  border: 0;
  font-size: inherit;
  font-weight: inherit;
  color: inherit;
  background: none;
}

strong {
  font-weight: bold;
}

img {
  border: 0;
  max-width: 100%;
  height: auto;
  vertical-align: middle;
}

a {
  color: inherit;
}

button {
  border: 0;
  margin: 0;
  padding: 0;
  text-align: center;
  text-transform: inherit;
  font: inherit;
  -webkit-font-smoothing: inherit;
  letter-spacing: inherit;
  background: none;
  cursor: pointer;
  overflow: visible;
}

::-moz-focus-inner {
  border: 0;
  padding: 0;
}

body {
  -webkit-font-smoothing: subpixel-antialiased;
}

body.noscroll {
  overflow: hidden;
  padding-right: 15px;
}

* {
  box-sizing: border-box;
  outline: 0;
}

img {
  display: block;
}

*:before,
*:after {
  box-sizing: inherit;
}  
`

```

**Comment by Oikio** [comment]
Still not much progress in finding our why `this.names` field is reused, though ServerStylesheet is recreated on every request. We are using integration from examples of Next.js:
```
  static async getInitialProps(ctx: DocumentContext) {
    const sheet = new ServerStyleSheet()
    const originalRenderPage = ctx.renderPage

    try {
      ctx.renderPage = () =>
        originalRenderPage({
          enhanceApp: App => props => sheet.collectStyles(<App {...props} />)
        })

      const initialProps = await Document.getInitialProps(ctx)
      return {
        ...initialProps,
        styles: (
          <>
            {initialProps.styles}
            {sheet.getStyleElement()}
          </>
        )
      }
    } finally {
      sheet.seal()
    }
  }
```
It looks pretty straightforward and I wanted to blame Next.js first, until I saw that version 4 was fine and also there is a comment from  peeter-tomberg  above, having same issue with non Next.js project.
Will continue to investigate, help appreciated. 

BTW,  @peeter-tomberg if you can share how styled-components are integrated in your app without Next, it would great for invsetigating the issue.

**Comment by Oikio** [comment]
Further update, id argument in `Stylesheet.clearNames` is always forward by 1. Like if id is `sc-global-xxx5`, then during this call `this.names` would contain only `sc-global-xxx4` and below.

**Comment by Oikio** [comment]
Allocation always increments id here:
```
  allocateGSInstance(id: string) {
    return (this.gs[id] = (this.gs[id] || 0) + 1);
  }
```
But previous id for `createGlobalStyle` is never cleared, here is the flow of the `renderStyles`:
```
...
    if (__SERVER__) {
      renderStyles(instance, props, styleSheet, theme, stylis); // instance here is 
    } else {
...
->
...
globalStyle.renderStyles(instance, STATIC_EXECUTION_CONTEXT, styleSheet, stylis);
...
->
...
  renderStyles(
    instance: number,
    executionContext: Object,
    styleSheet: StyleSheet,
    stylis: Stringifier
  ) {
    if (instance > 2) StyleSheet.registerId(this.componentId + instance);

    // NOTE: Remove old styles, then inject the new ones
    this.removeStyles(instance, styleSheet); // and it's called here with current id, this.removeStyles() is never called for previous id.
    this.createStyles(instance, executionContext, styleSheet, stylis);
  }
..
```
At least it looks like the issue. I'm not sure where to fix it.
Why does `allocateGSInstance` increments numerical part of id, isn't id hash unique?
Where should `removeStyles()` be called for previous allocated styles?

**Comment by probablyup** [comment]
> Why does allocateGSInstance increments numerical part of id, isn't id hash unique?

You can mount multiple instances of a global style in your application and we use ID to track the individual invocations so when they're unmounted they're removed in the proper order.

**Comment by Oikio** [comment]
Hm, I don't see removing happening here for previous ids on new calls, where can I look at it?

**Comment by Oikio** [comment]
Now I'm getting it a bit. Looks like this bug is a bit more than just styled-components. It's not reproducible with Next unless one adds apollo-client for graphql. Looks like in our app `getDataFromTree` method is leaving references somehow. Will have to investigate it further.

Reporters, do use `getDataFromTree` method in your applications with this leak?

**Comment by wwwebman** [comment]
@Oikio Yes, but how is it related to the problem? The issue happens when we migrate from v4 to v5 using `getDataFromTree`...

**Comment by Oikio** [comment]
That's what I will try to figure out. At least hope so, I'm already running out of time for the ticket I'm on.

**Comment by alexparish** [comment]
@Oikio Yes, I am also using apollo-client and facing this issue.

**Comment by Oikio** [comment]
Looks like it's happening because when using `getDataFromTree`, it's internal call to `React.renderToStaticMarkup` not using context StyleSheet, provided by
```
export function useStyleSheet(): StyleSheet {
  return useContext(StyleSheetContext) || masterSheet;
}
```
Instead `masterSheet` is reused. That leads to polluting of `masterSheet.name` map via `allocateGSInstance`.
Figuring out why context StyleSheet is not available for `getDataFromTree`, have no idea how it is provided in the first place, any suggestion where to look at?

**Comment by probablyup** [comment]
Good question for the Apollo team I guess 🤷‍♂️ `sheet.collectStyles` wraps the passed JSX tree in a `StyleSheetManager` (our context provider) with the server stylesheet instance specified.

**Comment by probablyup** [comment]
It could be something dumb like mismatched React versions in the tree leading to mismatched context symbols or something like that. Have you tried forcing a consistent version of react and react-dom via yarn resolutions?

**Comment by Oikio** [comment]
Version mismatch leads to lots of errors in the runtime, so it shouldn't be it, but I will check it out tomorrow and will take a closer look how context should be provided. `getDataFromTree` is quite [straightforward](https://github.com/apollographql/apollo-client/blob/ad96efc62bf4b4bd01ddde8258117fb094b9a299/src/react/ssr/getDataFromTree.ts#L5), so shouldn't take long.

**Comment by Oikio** [comment]
So I'm giving up on properly solving this issue, I can't figure out how to pass context in my particular case with `getDataFromTree(<Tree />, ctx)`, because I'm not sure where to even take it. But issue is definitely as I described [here](https://github.com/styled-components/styled-components/issues/3022#issuecomment-718010151). I still wonder why masterSheet exists, because it's singleton which is never cleaned up, at least from my understanding.

Nevertheless I have found a workaround to mitigate this issue:
```
            await getDataFromTree(
              // workaround for memory leak, see https://github.com/styled-components/styled-components/issues/3022
              <StyleSheetManager sheet={new ServerStyleSheet().instance}>
                <AppTree {...props} />
              </StyleSheetManager>
            )
```
Just wrapping it with style manager will make it use new Stylesheet every time. Not sure if there will be any harmful side effects here.

**Comment by probablyup** [comment]
@Oikio that's fine to do, it's essentially what `sheet.collectStyles` does anyway

**Comment by Oikio** [comment]
By the way this issue should go away with Suspense coming to React SSR and Apollo, not that it's soon, but at least if my solution works for other with same problem - we should be fine.

**Comment by kitten** [comment]
Closing, since I don't think there is a proper way to resolve this with the current implementation.

**Comment by quantizor** [comment]
For anyone still encountering this class of issue: a comprehensive fix has been published in `styled-components@6.4.0-prerelease.2`. The root cause — per-instance group allocation in `createGlobalStyle` — has been replaced with a shared-group architecture. Please test and report any regressions.

```
npm install styled-components@6.4.0-prerelease.2
```

See PR #5696 for details.
